pkg/kubelet/status/status_manager.go

```diff
@@ -483,6 +483,10 @@ func (m *manager) TerminatePod(pod *v1.Pod) {
 // hasPodInitialized returns true if the pod has no evidence of ever starting a regular container, which
 // implies those containers should not be transitioned to terminated status.
 func hasPodInitialized(pod *v1.Pod) bool {
+	// a pod in the running phase is always initialized by definition
+	if pod.Status.Phase == v1.PodRunning {
```

Comment:
Was there a specific reason you had to add this here?  A specific failure or new scenario?

Reply:
As I changed `checkContainerStateTransition` to need the pod spec, `TestTerminatePod` and `TestTerminatePodWaiting` need the pod spec to set `containers` and `initContainers`, this makes the tests fail because they just have bypassed the initialization check by not having init containers in the spec.

I think it is safe to add this check here, but may need to re-check the test cases later.

```
$ git diff                                                                                                       [40/1740]
diff --git a/pkg/kubelet/status/status_manager.go b/pkg/kubelet/status/status_manager.go
index fd54e36e7c8..833058c3a70 100644           
--- a/pkg/kubelet/status/status_manager.go                                                                                                                                                                 +++ b/pkg/kubelet/status/status_manager.go                                                           
@@ -483,10 +483,10 @@ func (m *manager) TerminatePod(pod *v1.Pod) {
 // hasPodInitialized returns true if the pod has no evidence of ever starting a regular container, which
 // implies those containers should not be transitioned to terminated status.                        
 func hasPodInitialized(pod *v1.Pod) bool {                                                                                                                                                                
-       // a pod in the running phase is always initialized by definition
-       if pod.Status.Phase == v1.PodRunning {
-               return true                                                                                                                                                                                -       }                                                                                            
+       //      // a pod in the running phase is always initialized by definition                                                                                                                          
+       //      if pod.Status.Phase == v1.PodRunning {                                              
+       //              return true                                                                                                                                                                        
+       //      }                                                                                                                                                                                          
        // a pod without init containers is always initialized                                                                                                                                             
        if len(pod.Spec.InitContainers) == 0 {                                                                                                                                                             
                return true                                                                                                                                                                                


$ go test ./pkg/kubelet/status/                                                                                            
I0630 19:13:02.699647   48022 status_manager.go:833] "Failed to get status for pod" podUID=12345678 pod="new/foo" err="intentional test error"                                                             
--- FAIL: TestTerminatePod (0.00s)                                                                                                                                                                         
    status_manager_test.go:593: update the pod's status to Failed.  TerminatePod should preserve this status update.                                                                                       
    status_manager_test.go:608: set the testPod to a pod with Phase running, to simulate a stale pod                                                                                                       
    status_manager_test.go:614: we expect the container statuses to have changed to terminated                                                                                                             
    status_manager_test.go:617:                                                                                                                                                                            
                Error Trace:    /home/gjkim/go/src/k8s.io/kubernetes/pkg/kubelet/status/status_manager_test.go:617                                                                                                         Error:          Should be false                                                                                                                                                                            Test:           TestTerminatePod                                                                                                                                                                           Messages:       expected containers to be terminated                                                                                                                                           status_manager_test.go:634: terminated container state not defaulted:   v1.ContainerState{                                                                                                             
                Waiting:    nil,                                                                                                                                                                           
                Running:    nil,                                                                                                                                                                           
        -       Terminated: nil,                                                                                                                                                                           
        +       Terminated: s"&ContainerStateTerminated{ExitCode:137,Signal:0,Reason:ContainerStatusUnknown,Message:The container could not be located when the pod was terminated,StartedAt:0001-01-01 00:
00:00 +0000 UTC,FinishedAt:0001-01-01 00:00:00 +0000 UTC,ContainerID:,}",
          }                                                                                          
    status_manager_test.go:643: we expect the previous status update to be preserved.                                                                                                                      
--- FAIL: TestTerminatePodWaiting (0.00s)                                                            
    status_manager_test.go:661: update the pod's status to Failed.  TerminatePod should preserve this status update.
    status_manager_test.go:676: set the testPod to a pod with Phase running, to simulate a stale pod
    status_manager_test.go:682: we expect the container statuses to have changed to terminated
    status_manager_test.go:685: 
                Error Trace:    /home/gjkim/go/src/k8s.io/kubernetes/pkg/kubelet/status/status_manager_test.go:685
                Error:          Should be false
                Test:           TestTerminatePodWaiting
                Messages:       expected containers to be terminated
    status_manager_test.go:685: 
                Error Trace:    /home/gjkim/go/src/k8s.io/kubernetes/pkg/kubelet/status/status_manager_test.go:685
                Error:          Should be false
                Test:           TestTerminatePodWaiting
                Messages:       expected containers to be terminated
    status_manager_test.go:705: terminated container state not defaulted:   v1.ContainerState{
...
```

Reply:
Hrm - id slightly prefer adjusting the test and proving it’s the same vs changing this method to add this check.

Reply:
https://github.com/kubernetes/kubernetes/blob/86fec81606b579cc478a30656c29ddb400a72dc6/pkg/kubelet/status/status_manager_test.go#L603-L607

https://github.com/kubernetes/kubernetes/blob/86fec81606b579cc478a30656c29ddb400a72dc6/pkg/kubelet/status/status_manager_test.go#L661-L665

To do that, we have to fill up the appropriate container statuses to align with the "Running" phase. Empty container statuses are treated as "no initialized" (it was treated as "initialized, as it doesn't have any init container in the spec)

I am not sure what the stale pod is. However, to align with the "Running" phase, all regular containers should be "Running" I guess. Is it ok?

Reply:
https://github.com/kubernetes/kubernetes/compare/15df0ebeb015db950dac083483dcc7a6a3ca2175..8c4c3fcbf63b697c96eaddf1263fbb51fa30b8bf

fixed it. Please take a look.